From a4c121a2774d2f5fc72b54642ee221841ab275a7 Mon Sep 17 00:00:00 2001
From: Ilya Ledvich <ilya@compulab.co.il>
Date: Sun, 25 Sep 2016 11:33:52 +0300
Subject: [PATCH 6/8] arm: imx7d: cl-som-imx7: add support for board specific
 device tree adjustment

Implement board specific routine fdt_board_adjust() for adjusting
Device Tree Blob (.dtb) before booting kernel.

Signed-off-by: Ilya Ledvich <ilya@compulab.co.il>
---
 board/compulab/cl_som_imx7/cl_som_imx7.c | 34 ++++++++++++++++++++++++++++++++
 configs/cl_som_imx7_defconfig            |  1 +
 2 files changed, 35 insertions(+)

diff --git a/board/compulab/cl_som_imx7/cl_som_imx7.c b/board/compulab/cl_som_imx7/cl_som_imx7.c
index 0632931..ae844a8 100644
--- a/board/compulab/cl_som_imx7/cl_som_imx7.c
+++ b/board/compulab/cl_som_imx7/cl_som_imx7.c
@@ -511,3 +511,37 @@ int checkboard(void)
 	return 0;
 }
 
+#ifdef CONFIG_OF_BOARD_SETUP
+#include <malloc.h>
+#include "../common/common.h"
+
+int fdt_board_adjust(void)
+{
+	u32 cpurev = get_cpu_rev();
+
+	/* Disable features not supported by i.MX7Solo */
+	if (((cpurev & 0xFF000) >> 12) == MXC_CPU_MX7S) {
+		/* FEC2 with PHY */
+		fdt_node_disable("/soc/aips-bus@30800000/ethernet@30bf0000");
+		fdt_node_disable("/soc/aips-bus@30800000/ethernet@30be0000/mdio/ethernet-phy@1");
+		/* PCIe */
+		fdt_node_disable("/soc/pcie@0x33800000");
+		/* USB Host HSIC */
+		fdt_node_disable("/soc/aips-bus@30800000/usb@30b20000");
+	}
+
+	/* Main storage setup */
+	if (nand_enabled) {
+		/* Enable GPMI and disable eMMC */
+		fdt_node_enable("/soc/gpmi-nand@33002000");
+		fdt_node_disable("/soc/aips-bus@30800000/usdhc@30b60000");
+	} else {
+		/* Enable eMMC and disable GPMI */
+		fdt_node_enable("/soc/aips-bus@30800000/usdhc@30b60000");
+		fdt_node_disable("/soc/gpmi-nand@33002000");
+	}
+
+	return 0;
+}
+#endif
+
diff --git a/configs/cl_som_imx7_defconfig b/configs/cl_som_imx7_defconfig
index c0c0b56..d9776b6 100644
--- a/configs/cl_som_imx7_defconfig
+++ b/configs/cl_som_imx7_defconfig
@@ -4,6 +4,7 @@ CONFIG_TARGET_CL_SOM_IMX7=y
 CONFIG_ARMV7_BOOT_SEC_DEFAULT=y
 CONFIG_IMX_RDC=y
 CONFIG_IMX_BOOTAUX=y
+CONFIG_OF_BOARD_SETUP=y
 CONFIG_SYS_EXTRA_OPTIONS="IMX_CONFIG=board/compulab/cl_som_imx7/imximage.cfg"
 CONFIG_SPI_BOOT=y
 CONFIG_BOOTDELAY=3
-- 
1.9.1

